gnucash/import-export/import-main-matcher.c

```diff
@@ -921,81 +1022,130 @@ gnc_gen_trans_edit_fields (GtkMenuItem *menuitem, GNCImportMainMatcher *info,
         return;
     }
 
-    if (selected_rows->next)
-    {
-        LEAVE ("User selected multiple rows, not supported");
-        return;
-    }
-
-    g_return_if_fail (gtk_tree_model_get_iter (model, &iter,
-                                               selected_rows->data));
+    row_info_list = gnc_g_list_map (selected_rows, (GncGMapFunc) row_get_info, info);
 
-    gtk_tree_model_get (model, &iter, DOWNLOADED_COL_DATA,
-                        &trans_info, -1);
-    trans = gnc_import_TransInfo_get_trans (trans_info);
+    desc_hash = g_hash_table_new (g_str_hash, g_str_equal);
+    notes_hash = g_hash_table_new (g_str_hash, g_str_equal);
+    memo_hash = g_hash_table_new (g_str_hash, g_str_equal);
 
-    switch (field)
+    for (GList *n = row_info_list; n; n = n->next)
     {
-        case DESCRIPTION:
-        {
-            char* new_field =
-                gnc_input_dialog_with_entry(info->main_widget, "",
-                                            _("Enter new Description"),
-                                            xaccTransGetDescription (trans));
-            if (!new_field) break;
-            xaccTransSetDescription (trans, new_field);
-            gtk_tree_store_set (store, &iter, DOWNLOADED_COL_DESCRIPTION,
-                                new_field, -1);
-            g_free (new_field);
-            break;
-        }
-        case MEMO:
+        const RowInfo *info = n->data;
+        Account *acc = xaccSplitGetAccount (info->split);
+        if (g_list_find (accounts_scanned, acc))
+            continue;
+
+        accounts_scanned = g_list_prepend (accounts_scanned, acc);
+        for (GList *n = xaccAccountGetSplitList (acc); n; n = n->next)
         {
-            Split *first_split =
-                gnc_import_TransInfo_get_fsplit (trans_info);
-            char *new_field =
-                gnc_input_dialog_with_entry(info->main_widget, "",
-                                            _("Enter new Memo"),
-                                            xaccSplitGetMemo (first_split));
-            if (!new_field) break;
-            xaccSplitSetMemo (first_split, new_field);
-            gtk_tree_store_set (store, &iter,
-                                DOWNLOADED_COL_MEMO, new_field, -1);
-            g_free (new_field);
-            break;
+            const Split *s = n->data;
+            const Transaction *t = xaccSplitGetParent (s);
+            const gpointer one = GINT_TO_POINTER (1);
+            const gchar *key;
+
+            key = xaccTransGetDescription (t);
+            if (key && *key)
+                g_hash_table_insert (desc_hash, (gpointer)key, one);
+
+            key = xaccTransGetNotes (t);
+            if (key && *key)
+                g_hash_table_insert (notes_hash, (gpointer)key, one);
+
+            key = xaccSplitGetMemo (s);
+            if (key && *key)
+                g_hash_table_insert (memo_hash, (gpointer)key, one);
```

Comment:
Separate this out and do it as part of the matcher setup and add a hash entry if the user creates a new description/note/memo.

While you're at it create a struct with all three hash tables that you can pass around.

Reply:
> Separate this out and do it as part of the matcher setup and add a hash entry if the user creates a new description/note/memo.
> 
> While you're at it create a struct with all three hash tables that you can pass around.

Agree about doing in the matcher setup but do you mean creating a signal to watch for new transactions?

Reply:
Hmm, hadn't thought about that, I was thinking of using the dialog to create a new description/note/memo. Monitoring the QofEvents for `xaccTransCommitEdit` would be better, yes. I don't think you need to create anything, just catch and act on what's there.

Reply:
> Hmm, hadn't thought about that, I was thinking of using the dialog to create a new description/note/memo. Monitoring the QofEvents for `xaccTransCommitEdit` would be better, yes. I don't think you need to create anything, just catch and act on what's there.

Don't think I'll add a signal... no idea how it works. The import matcher does not have ability to add a new transaction while it's running, therefore I'm not sure it's necessary to update the desc/notes/memo hash tables during the matcher lifetime. The code currently written will populate GtkListStore with the desc/notes/memo with *existing* data, plus the current field contents. So, if you've renamed a desc from "INTEREST" to "NONE", and re-edit it, the autocomplete will include NONE rather than INTEREST. I think this is a reasonable approach.

Reply:
> if the user launches import matcher, clicks behind into the register, adds some more transactions, then the hash could be updated.

OK, probably a rare occurrence; it's more likely in the reconcile step. 

> So, if you've renamed a desc from "INTEREST" to "NONE", and re-edit it, the autocomplete will include NONE rather than INTEREST. I think this is a reasonable approach.

Erm, not quite. Suppose you have previous transactions with a description of "Interest" and you edit a batch of imported transactions' descriptions to "Interesting", a string that wasn't previously used. You now select a different batch of transactions and edit the descriptions. "Interesting" will not be in the list, and some user is going to file a bug saying that it should be.

Reply:
Ok see very last commit -- editing a field causes the new text to be added into the hash. This means the hash will keep growing during a matcher session; editing "Interest" to "Interesting" then re-edit to "Interested" means both are now suggestions, until they are reset in the next matcher session.

Reply:
> Ok see very last commit -- editing a field causes the new text to be added into the hash. This means the hash will keep growing during a matcher session; editing "Interest" to "Interesting" then re-edit to "Interested" means both are now suggestions, until they are reset in the next matcher session.

Oops doesn't quite work: The hash table keys are pointers to original string data, and an edited text points to garbage. Must be a rethink... Maybe a separate 'new_desc' hash.